plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/index/HudiRecordLevelIndexSupport.java

```diff
@@ -48,81 +57,87 @@ public class HudiRecordLevelIndexSupport
+                            timer.endTimer(), schemaTableName);
+                    return Optional.empty();
+                }
+                log.debug(String.format("Constructed %d record keys for index lookup.", recordKeys.size()));
+
+                // Perform index lookup in metadataTable
+                // TODO: document here what this map is keyed by
+                Map<String, HoodieRecordGlobalLocation> recordIndex = lazyTableMetadata.get().readRecordIndex(recordKeys);
+                if (recordIndex.isEmpty()) {
+                    log.debug("Record level index lookup took %s ms but returned no locations for the given keys %s for table %s",
+                            timer.endTimer(), recordKeys, schemaTableName);
+                    // Return all original fileSlices
+                    return Optional.empty();
+                }
+
+                // Collect fileIds for pruning
+                Set<String> relevantFiles = recordIndex.values().stream()
+                        .map(HoodieRecordGlobalLocation::getFileId)
+                        .collect(Collectors.toSet());
+                log.debug(String.format("Record level index lookup identified %d relevant file IDs.", relevantFiles.size()));
+
+                return Optional.of(relevantFiles);
+            });
         }
-        List<String> recordKeyFields = Arrays.asList(recordKeyFieldsOpt.get());
+        this.futureStartTimeMs = System.currentTimeMillis();
+    }
 
-        // Only extract the predicates relevant to the record key fields
-        TupleDomain<String> filteredDomains = extractPredicatesForColumns(regularColumnPredicates, recordKeyFields);
+    @Override
+    public boolean shouldSkipFileSlice(FileSlice slice)
+    {
+        try {
+            if (relevantFileIdsFuture.isDone()) {
+                Optional<Set<String>> relevantFileIds = relevantFileIdsFuture.get();
+                return relevantFileIds.map(fileIds -> !fileIds.contains(slice.getFileId())).orElse(false);
+            }
 
-        // Construct the actual record keys based on the filtered predicates using Hudi's encoding scheme
-        List<String> recordKeys = constructRecordKeys(filteredDomains, recordKeyFields);
+            long elapsedMs = System.currentTimeMillis() - futureStartTimeMs;
+            if (elapsedMs > recordIndexWaitTimeout.toMillis()) {
+                // Took too long; skip decision
+                return false;
+            }
 
-        if (recordKeys.isEmpty()) {
-            // If key construction fails (e.g., incompatible predicates not caught by canApply, or placeholder issue)
-            log.warn("Took %s ms, but could not construct record keys from predicates. Skipping record index pruning for table %s",
-                    timer.endTimer(), schemaTableName);
-            return inputFileSlices;
+            long remainingMs = recordIndexWaitTimeout.toMillis() - elapsedMs;
```

Comment:
nit: for all index support, this can be simplified as `Math.max(0, recordIndexWaitTimeout.toMillis() - elapsedMs)`?